### DISCUSSION TITLE
How to have Enums with values

### DISCUSSION THREAD

**Discussion by yashrathi-git** [pydantic/pydantic#3068] [Question] [answered]
How can I use Enum to represent choice and different values for those choices?  What I mean is when I initialize my Pydantic model, I want to pass in one of valid choice as **string**, but when I access that value (with `MyModel.field.value`), I want it to return a callable according to the string that was provided.
```py
from enum import Enum

from pydantic import BaseModel, ValidationError


class CallableEnum(Enum):
    # I want if `apple` is provided in type str, while accessing `Model.enum_field.value`, it gives me the callable.
    apple = lambda x: print(x)        # This callable just a useless placeholder

class Model(BaseModel):
    field: CallableEnum

model = Model(field = "apple")    
model.apple    # I want this to return the callable

```
If this could be done, it would make my code, lot more cleaner.

Thank You

**Comment by PrettyWood** [comment, marked as answer]
Hi @yashrathi-git 
This has nothing to do with pydantic but stdlib enums
```py
from enum import Enum

class C(Enum):
    a = lambda x: x
    b = "f"

print(C.__members__)
# {'b': <C.b: 'f'>}
```
So how do you do? I think the easiest is just to wrap you function in a class
```py
from enum import Enum

from pydantic import BaseModel


class CallableWrapper:
    def __init__(self, f):
        self.f = f

    def __call__(self, *args, **kwargs):
        return self.f(*args, **kwargs)


class CallableEnum(Enum):
    apple = CallableWrapper(lambda x: print(x))

class Model(BaseModel):
    field: CallableEnum

    class Config:
        use_enum_values = True

model = Model(field=CallableEnum.apple)
model.field("it works!")
```

Hope it helps

**Reply by yashrathi-git** [reply]
```py
from enum import Enum

from pydantic import BaseModel, ValidationError


class C(Enum):
    a = lambda x: x
    b = "f"

class Model(BaseModel):
    field: C

model = Model(field = "a")     # This returns `ValidationError` saying it a is not an acceptable choice
""""
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "pydantic/main.py", line 406, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Model
c
  value is not a valid enumeration member; permitted: 'f' (type=type_error.enum; enum_values=[<C.b: 'f'>])
""""
```
@PrettyWood How could I make it work, so it takes the choice str from attribute name, and value from their values?

**Reply by PrettyWood** [reply]
@yashrathi-git Have you actually read my answer? And btw a hi and a thanks never hurt when asking people for help

**Reply by yashrathi-git** [reply]
@PrettyWood Hi, I really appreciate your help. Thanks a lot :)
Sorry, I don't know what happened, I am probably blind or something for sure (I think I started typing reply before the most recent edit, and Github didn't sync it without refresh). I wasn't able to see the last part. Again I really appreciate you taking the time to help me. And sorry, for being rude.

**Reply by PrettyWood** [reply]
Sorry for overreacting. I'm glad it worked.
